src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -2599,37 +2600,36 @@ internal override ImmutableArray<TypeWithAnnotations> TypeArgumentsWithAnnotatio
                 }
             }
 
-            public override ImmutableArray<TypeParameterSymbol> TypeParameters
-            {
-                get
-                {
-                    EnsureTypeParametersAreLoaded();
-                    return _lazyTypeParameters;
-                }
-            }
+            public override ImmutableArray<TypeParameterSymbol> TypeParameters => EnsureTypeParametersAreLoaded();
 
             internal sealed override NamedTypeSymbol AsNativeInteger() => throw ExceptionUtilities.Unreachable();
 
             internal sealed override NamedTypeSymbol NativeIntegerUnderlyingType => null;
 
-            private void EnsureTypeParametersAreLoaded()
+            private ImmutableArray<TypeParameterSymbol> EnsureTypeParametersAreLoaded()
             {
-                if (_lazyTypeParameters.IsDefault)
-                {
-                    var moduleSymbol = ContainingPEModule;
+                return InterlockedOperations.Initialize(
+                    ref _lazyTypeParameters,
+                    static self =>
+                    {
+                        if (self._arity == 0)
```

Comment:
> if (self._arity == 0) [](http://example.com/codeflow?start=23&length=22)

Is this the optimization we want to add? Then all the other rewrites in this file are not necessary in my opinion. I suggest reverting them. Basically, I would expect no changes in ```TypeParameters``` property, no lambdas and signature changes in ```EnsureTypeParametersAreLoaded```,  switch to ```ArrayBuilder``` would be fine though.

Reply:
@AlekseyTs This was the suggestion from @sharwell. https://github.com/dotnet/roslyn/pull/68978#discussion_r1260144847

Reply:
> This was the suggestion

I still think that the churn isn't necessary. Please revert as suggested.